destination/writer/writer.go

```diff
@@ -0,0 +1,317 @@
+// Copyright © 2022 Meroxa, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package writer
+
+import (
+	"context"
+	"database/sql"
+	"encoding/json"
+	"fmt"
+	"reflect"
+	"strings"
+
+	"github.com/conduitio-labs/conduit-connector-oracle/config"
+	sdk "github.com/conduitio/conduit-connector-sdk"
+	"github.com/huandu/go-sqlbuilder"
+)
+
+const (
+	// metadata related.
+	metadataTable  = "table"
+	metadataAction = "action"
+
+	// action names.
+	actionUpdate = "update"
+	actionDelete = "delete"
+)
+
+// Writer implements a writer logic for Oracle destination.
+type Writer struct {
+	db  *sql.DB
+	cfg config.Destination
+}
+
+// New creates new instance of the Writer.
+func New(db *sql.DB, cfg config.Destination) *Writer {
+	return &Writer{
+		db:  db,
+		cfg: cfg,
+	}
+}
+
+// Write writes a sdk.Record into a Destination.
+func (w *Writer) Write(ctx context.Context, record sdk.Record) error {
+	switch record.Metadata[metadataAction] {
+	case actionUpdate:
+		return w.update(ctx, record)
+	case actionDelete:
+		return w.delete(ctx, record)
+	default:
+		return w.insert(ctx, record)
+	}
+}
+
+// Close closes the underlying db connection.
+func (w *Writer) Close(ctx context.Context) error {
+	return w.db.Close()
+}
+
+// inserts a record.
+func (w *Writer) insert(ctx context.Context, record sdk.Record) error {
+	tableName := w.getTableName(record.Metadata)
+
+	payload, err := w.structurizeData(record.Payload)
+	if err != nil {
+		return fmt.Errorf("structurize payload: %w", err)
+	}
+
+	// if payload is empty return empty payload error
+	if payload == nil {
+		return errEmptyPayload
+	}
+
+	columns, values := w.extractColumnsAndValues(payload)
+
+	w.convertValues(values)
+
+	query, err := w.buildInsertQuery(tableName, columns, values)
```

Comment:
What about primary keys?
Could we implement upsert?

Reply:
Do we need a primary key for inserting?
Oracle does not support an `upsert` operation.

Reply:
Oracle has merge function, we can try to use it like upsert
https://docs.oracle.com/database/121/SQLRF/statements_9017.htm#SQLRF01606

We have to handle case with duplication insert  uniq primary key error, for this case better to use merge

Reply:
I see.
Ok, I'll try to implement it.

Thanks!

Reply:
I've implemented `upsert` method instead of `insert` and `update`.